plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Cse.hs

```diff
@@ -0,0 +1,398 @@
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TupleSections     #-}
+{-# LANGUAGE TypeApplications  #-}
+{-# LANGUAGE TypeFamilies      #-}
+{-# LANGUAGE TypeOperators     #-}
+
+module UntypedPlutusCore.Transform.Cse (cse) where
+
+import PlutusCore (MonadQuote, Name, freshName)
+import UntypedPlutusCore.Core
+import UntypedPlutusCore.Purity (isWorkFree)
+import UntypedPlutusCore.Size
+
+import Control.Lens
+import Control.Monad
+import Control.Monad.Trans.Class
+import Control.Monad.Trans.Reader
+import Control.Monad.Trans.State.Strict
+import Data.Foldable
+import Data.Hashable
+import Data.HashMap.Strict (HashMap)
+import Data.HashMap.Strict qualified as Map
+import Data.List.Extra
+import Data.Traversable
+
+{- Note [CSE]
+
+-------------------------------------------------------------------------------
+1. Simplifications
+-------------------------------------------------------------------------------
+
+This is a simplified (i.e., not fully optimal) implementation of CSE. The two simplifications
+we made are:
+
+- No alpha equivalence check, i.e., `\x -> x` and `\y -> y` are considered different expressions.
+- The builtin function arity information is approximate: rather than using the accurate arities,
+  we simply use the maximum number of arguments applied to a builtin function in the program
+  as the builtin function's arity. The arity information is used to determine whether a builtin
+  application is possibly saturated.
+
+-------------------------------------------------------------------------------
+2. How does it work?
+-------------------------------------------------------------------------------
+
+We use the following example to explain how the implementation works:
+
+\x y -> (1+(2+x))
+        +
+        (case y [ (1+(2+x)) + (3+x)
+                , (2+x) + (3+x)
+                , 4+x
+                ]
+        )
+
+The implementation makes several passes on the given term. The first pass collects builtin
+arity information as described above.
+
+In the second pass, we assign a unique ID to each `LamAbs`, `Delay`, and each `Case` branch.
+Then, we annotate each subterm with a path, consisting of IDs encountered from the root
+to that subterm (not including itself).
+
+In the above example, the ID of `\x` is 0, the ID of `\y` is 1, and the IDs of the
+three case branches are 2, 3, 4 (the actual numbers don't matter, as long as they are unique).
+The path for the first `1+(2+x)` and the first `2+x` is "0.1"; the path for the second
+`1+(2+x)` and the second `2+x` is "0.1.2"; the path for `4+x` is "0.1.4".
+
+In the third pass, we calculate a count for each `(term, path)` pair, where `term` is a
+non-workfree term, and `path` is its path. If the same term has two paths, and one is an
+ancestor (i.e., prefix) of the other, we increment the count for the ancestor path in both
+instances.
```

Comment:
Because we need to aggregate the counts on each path via `Map.fromListWith`.

Reply:
Sorry, still don't get it. Can you have a map from hashes of terms to lists of paths, without even any counts, and then post-process each list of paths? If you have `2+x :-> ["0.1", "0.1.2", "0.1.3"]`, then during post-processing you figure out that `"0.1"` is the root and insert a binding at that point, ignoring all the path extensions. Technically, it's not even a list of paths as a value, it's a trie of paths and "post-processing" means simply taking the root of each forest at the beginning of the trie (unless the forest consists only of the root, i.e. the subexpression only occurs once).

Reply:
Here's an example. Suppose for `2+x`, we currently have the following counts: 
```
"0.1.2" -> 10, "0.1.3" -> 20, "0.2.2" -> 30
```

Now we process path "0.1". After processing it, the counts will become: 
```
"0.1" -> 10, "0.1" -> 20, "0.2.2" -> 30`, "0.1" -> 1
```

Finally we use `Map.fromListWith` to aggregate them, and get

```
"0.1" -> 31, "0.2.2" -> 30
```

Does that make sense?

Reply:
Actually paths indeed don't need to be map keys... I'll update the code

Reply:
The inner HashMap (with Path as key) is replaced with a list in b32cd3cc99f9fd31b1cc0e52b838d050cedc4850, because in my example above, we can easily go from step 1 directly to step 3, bypassing step 2.